lib/srv/alpnproxy/auth/auth_proxy.go

```diff
@@ -113,13 +114,21 @@ func (s *AuthProxyDialerService) dialLocalAuthServer(ctx context.Context) (net.C
 	if len(authServers) == 0 {
 		return nil, trace.NotFound("empty auth servers list")
 	}
-	//TODO(smallinksy) Better support for HA. Add dial retry on auth network errors.
-	authServerIndex := rand.Intn(len(authServers))
-	conn, err := net.Dial("tcp", authServers[authServerIndex].GetAddr())
-	if err != nil {
-		return nil, trace.Wrap(err)
+	var errors []string
+
+	//select random server from list, try others if it's unavailable
+	for len(authServers) > 0 {
```

Comment:
Instead of doing this, just build a map with the auth servers (or even just the addresses) as keys, and iterate over it; iteration order is random (a little biased, but it's going to be biased in hash space anyway).

Reply:
My 2c: I'd rather take the random, otherwise I fear this becomes stable if the list of auth servers is always the same.

Reply:
I'd rather not do that for couple of reasons:

- iteration order is currently starting from random start point but then it's always the same, so for map {1:1,2:2,3:3} iteration order can be (keys only): [1,2,3], [2,3,1] or [3,1,2]. If first server is unavailable, then second one will handle 2/3 of the traffic instead of 50%
- iteration order is undefined not random so it can change in future versions of Go (unlikely but still)
- most of the time first server will work so building map will be waste of time

Reply:
You're right, for a small fixed list turned to map (which is what we'll almost always have) the iteration order is not only the same except rotated, but it's also very much skewed towards a specific order.

It's fine like this although I'd prefer swapping in the last item in the slice in the removed item's position instead of sliding everything by one index.

Reply:
That's even better, I've changed it